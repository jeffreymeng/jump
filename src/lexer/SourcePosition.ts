/**
 * A SourcePosition refers to a specific position in the source.
 * Each character in the source has a unique position.
 */
export default class SourcePosition {
	/**
	 * @param source - The text of the source code.
	 * @param index - The index of the character within the source text.
	 */
	constructor(protected readonly source: string, protected readonly index: number) {
		if (index >= source.length) {
			throw new Error("SourcePosition expected an index less than the length of source.")
		}
	}

	/**
	 * Returns the line of the index in the given source text. Lines are 1-indexed.
	 */
	public getLine(): number {
		return this.source.substring(0, this.index).split("\n").length
	}

	/**
	 * Returns the index of the character within its line. Columns are
	 * 1-indexed.
	 */
	public getCol(): number {
		// given the column, look backwards for the closest newline strictly
		// before the current index, since if the current character is a
		// newline that doesn't count.
		return this.index - this.source.lastIndexOf("\n", this.index - 1);
	}

}
