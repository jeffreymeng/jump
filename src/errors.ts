import SourcePosition from "./lexer/SourcePosition";

export class JumpError extends Error {
	constructor(message: string) {
		super(message);
	}
}

/**
 * An error in the syntax of the input.
 */
export class JumpSyntaxError extends JumpError {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.position = position;
	}
}

export class JumpInternalError extends JumpError {
	constructor(message: string) {
		super(message);
	}
}

export class JumpDivisionByZeroError extends JumpError {
	constructor(dividend: number, public readonly position: SourcePosition) {
		super(`JumpDivisionByZeroError: Cannot divide ${dividend} by zero.`);
	}
}

/**
 * An error in the implementation of Jump's AST.
 */
export class JumpASTInternalError extends JumpError {
	constructor(message: string) {
		super(message);
	}
}
