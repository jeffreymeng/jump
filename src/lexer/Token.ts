import SourcePosition from "./SourcePosition";

export enum TokenType {
	STRING_LITERAL = "string_literal",
	INT_LITERAL = "int_literal",
	DOUBLE_LITERAL = "double_literal",
	OPERATOR = "operator",
	IDENTIFIER = "identifier",
	CONTROL = "control",
	KEYWORD = "keyword",
}
export default class Token {
	/**
	 *
	 * @param type - The type of the token
	 * @param symbol - The value of the token
	 */
	constructor(
		public readonly type: TokenType,
		public readonly symbol: string,
		public readonly position: SourcePosition
	) {}

	/**
	 * Returns whether or not the token matches another token, in terms of content,
	 * but not position.
	 * @param other
	 */
	public equals(other: Token) {
		return this.type === other.type && this.symbol === other.symbol;
	}

	/**
	 * Returns whether or not the token matches the provided type and, if provided, symbol.
	 * @param type
	 * @param symbol
	 */
	public is(type: TokenType, symbol?: string) {
		return this.type === type && (!symbol || this.symbol === symbol);
	}

	public toString() {
		return `Token<${this.type}, '${this.symbol}'>`;
	}
}
