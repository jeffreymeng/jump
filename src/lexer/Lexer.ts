import Token, { TOKEN_TYPE } from "./Token";
import SourcePosition from "./SourcePosition";
import Source from "./Source";

/**
 * An error that occurred while lexing, typically not due to the input but due to
 * an implementation bug.
 */
export class LexerError extends Error {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.name = "Lexer Error";
		this.position = position;
	}
}

/**
 * "SheetsScript" syntax error -- so it's not the same as a native syntax error.
 */
export class SSSyntaxError extends Error {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.name = "SSSyntax Error";
		this.position = position;
	}
}


export default class Lexer implements Iterator<Token> {
	/**
	 * The source code of the lexer.
	 */
	public readonly source: Source;

	/**
	 * The index of the current character within the source.
	 */
	protected index = 0;

	/**
	 * The index of the current token.
	 */
	protected currentTokenIndex = 0;

	constructor(source: Source) {
		this.source = source;
	}

	public static readonly DIGITS = "0123456789";



	/**
	 * Consumes the next token, and returns it.
	 */
	public nextToken(): Token {
		const c = this.source.next();
		this.setTokenIndex();

		if ("\"'".includes(c)) {
			// consume until we see the same type of quote again
			const buffer = this.source.consumeUntil(c);
			return new Token(TOKEN_TYPE.STRING, buffer);
		} else if (Lexer.DIGITS.includes(c)) {
			let buffer = c + this.source.consumeAll(Lexer.DIGITS);
			// if the next character is a dot, then this is part of a float, so keep going.
			if (this.source.peekNext() === ".") {
				buffer += this.source.next() + this.source.consumeAll(Lexer.DIGITS);
			}
			return new Token(TOKEN_TYPE.NUMBER, buffer);
		} else if (c === "\n") {
			return new Token(TOKEN_TYPE.CONTROL, "\n");
		} else if (/[ \t]/g.test(c)) {
			// skip any spaces and tabs.
			return this.nextToken();
		}

		throw new SSSyntaxError(
			`Unexpected character '${c}'.`,
			this.source.getPosition()
		);
	}

	public hasNextToken(): boolean {
		// if there are only whitespace characters (including newlines)
		// then we can
		return !/[\s\n]*/g.test(this.source.peekNext());
	}

	public next(): IteratorResult<Token> {
		if (!this.hasNextToken()) {
			return {
				value: null,
				done: true
			}
		}
		return {
			value: this.nextToken(),
			done: this.hasNextToken(),
		};
	}

	/**
	 * Set the current token position to the lexer's next character's position
	 */
	protected setTokenIndex() {
		this.currentTokenIndex = this.index;
	}
}
