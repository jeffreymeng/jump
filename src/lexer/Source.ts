import SourcePosition from "./SourcePosition";
import { JumpSyntaxError } from "./Lexer";

export default class Source {
	protected _index = 0;
	public readonly text: string;

	/**
	 * Returns the current index of the source.
	 */
	public get index() {
		return this._index;
	}

	constructor(text: string) {
		this.text = text;
	}

	/**
	 * Returns the next character from the source, updating the line and position as necessary.
	 * If no next character exists, returns an empty string.
	 */
	public next(): string {
		if (this._index >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of input.",
				new SourcePosition(this.text, this._index)
			);
		}
		const c = this.text[this._index];
		this._index++;
		return c;
	}

	/**
	 * Returns the next character from the source without changing the current position.
	 * @param offset - If provided, the ith character after the current one is returned.
	 * If `i` is out of range, an empty string will be returned.
	 */
	public peekNext(offset = 0): string {
		if (this._index + offset >= this.text.length) {
			return "";
		}

		return this.text[this._index + offset];
	}

	/**
	 * Returns the next `len` characters from the source without changing the current position.
	 * @param len - The number of characters to return, or `0` to return the rest of source. Must be in range
	 */
	public peekNextCharacters(len = 0): string {
		if (len === 0) {
			return this.text.substring(this._index);
		}

		if (this._index + len >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of file",
				new SourcePosition(this.text, this._index)
			);
		}
		return this.text.substring(this._index, this._index + len);
	}

	/**
	 * Consumes characters until any character inside endChars is found.
	 * The end character is consumed, but only the characters from
	 * the current position (inclusive) to the last character before
	 * the end character are returned. The string must eventually contain
	 * one of the endChars.
	 * @param endChars - A string containing one or more characters that will cause consumeUntil()
	 *  to stop consuming characters.
	 * @param errorMessageIfNotFound - If the string does not eventually contain
	 * an endChar, then this message will be used instead of the default error message.
	 */
	public consumeUntil(
		endChars: string,
		errorMessageIfNotFound = "Unexpected end of input."
	): string {
		const consumed = this.consumeWhile((c) => !endChars.includes(c));
		// consume the character that caused the predicate to fail, unless
		// we reached the end of the string already (and thus nothing caused
		// this to throw), which would cause an error.
		if (this._index < this.text.length) {
			this.next();
		} else {
			throw new JumpSyntaxError(
				errorMessageIfNotFound,
				this.getPosition()
			);
		}
		return consumed;
	}

	/**
	 * Consumes and returns characters until a character not in `chars` is found.
	 * @param chars - A string containing one or more characters to be consumed continuously.
	 */
	public consumeAll(chars: string): string {
		return this.consumeWhile((c) => chars.includes(c));
	}

	/**
	 * Consumes characters until the predicate is no longer true.
	 * All characters from the starting position until the predicate fails will be consumed and returned.
	 * The character that failed the predicate will NOT be consumed.
	 * @param predicate - A function returning true if a character passes the predicate.
	 */
	public consumeWhile(predicate: (c: string) => boolean): string {
		let buffer = "";
		while (this.hasNext() && predicate(this.peekNext())) {
			buffer += this.next();
		}
		return buffer;
	}

	/**
	 * Get a SourcePosition object corresponding to the `i`th character of the source,
	 * or the current index if none is provided.
	 */
	public getPosition(i = this._index) {
		return new SourcePosition(this.text, i);
	}

	/**
	 * Returns true if the source has a next token (e.g. has not reached the end yet).
	 */
	public hasNext() {
		return this._index < this.text.length;
	}
}
