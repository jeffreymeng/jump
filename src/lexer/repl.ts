import Token, { TOKEN_TYPE } from "./Token";
import Lexer from "./Lexer";
import Source from "./Source";
import * as readline from "readline";

// const s = `"hello" 2 0.2 42.294 "23.1"`;
// const expected = [
// 	new Token(TOKEN_TYPE.STRING_LITERAL, "hello"),
// 	new Token(TOKEN_TYPE.INT_LITERAL, "2"),
// 	new Token(TOKEN_TYPE.DOUBLE_LITERAL, "0.2"),
// 	new Token(TOKEN_TYPE.DOUBLE_LITERAL, "42.294"),
// 	new Token(TOKEN_TYPE.STRING_LITERAL, "23.1")
// ]



async function repl() {
	const rl = readline.createInterface({
		input: process.stdin,
		output: process.stdout
	});

	const ask = async (question: string): Promise<string> =>
		await new Promise((res, rej) =>
			rl.question(question, (answer) => res(answer)));



	console.log("Jump REPL v0.0.1")
	console.log("Type exit to exit.")
	while (true) {
		const input = await ask("Jump> ")
		if (input === "exit") break;

		try {
			const lexer = new Lexer(new Source(input));

			for (const token of lexer) {
				console.log(`Token<${token.type}, ${token.symbol}>`)
			}
		} catch (e) {
			console.log(e)
		}
	}

	rl.close();
}

repl()