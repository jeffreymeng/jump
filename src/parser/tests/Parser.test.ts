import { describe, expect, test } from "@jest/globals";
import { JumpSyntaxError } from "../../errors";
import execLine from "./execLine";
import Parser from "../Parser";
import Lexer from "../../lexer/Lexer";
import Source from "../../lexer/Source";
import { TokenType } from "../../lexer/Token";

class TestableParser extends Parser {
	public next(assertType?: TokenType, assertSymbol?: string) {
		if (assertType && assertSymbol) {
			return super.next(assertType, assertSymbol);
		} else if (assertType) {
			return super.next(assertType);
		} else {
			return super.next();
		}
	}
	public peek(n?: number) {
		return super.peek(n);
	}
	public peekMatches(type: TokenType, symbol?: string | string[]) {
		if (symbol) {
			return super.peekMatches(type, symbol);
		} else {
			return super.peekMatches(type);
		}
	}
}

describe("Internal next/peek/peekMatches", () => {
	test("Next", () => {
		const p = new TestableParser(new Lexer(new Source("1 2 3")));
		expect(p.next().is(TokenType.INT_LITERAL, "1")).toBe(true);
		expect(p.next().is(TokenType.INT_LITERAL, "2")).toBe(true);
	});

	test("Next Error Cases", () => {
		const p = new TestableParser(new Lexer(new Source("1 2 3")));
		expect(() => p.next(TokenType.INT_LITERAL, "1")).not.toThrow();
		expect(() =>
			p.next(TokenType.INT_LITERAL, "1")
		).toThrowErrorMatchingSnapshot();
		p.next();
		p.next();
		expect(() => p.next()).toThrowErrorMatchingSnapshot();
	});

	test("Peek", () => {
		const p = new TestableParser(new Lexer(new Source("1 2 3")));
		expect(p.peek()?.is(TokenType.INT_LITERAL, "1")).toBe(true);
		expect(p.peek()?.is(TokenType.INT_LITERAL, "1")).toBe(true);

		p.next();

		expect(p.peek()?.is(TokenType.INT_LITERAL, "2")).toBe(true);
		expect(p.peek(2)?.is(TokenType.INT_LITERAL, "3")).toBe(true);
		expect(p.peek()?.is(TokenType.INT_LITERAL, "2")).toBe(true);

		p.next(TokenType.INT_LITERAL, "2");
	});

	test("PeekMatches", () => {
		const p = new TestableParser(new Lexer(new Source("1 'hi'")));
		expect(p.peekMatches(TokenType.INT_LITERAL, "1")).toBe(true);
		expect(p.peekMatches(TokenType.INT_LITERAL, "2")).toBe(false);
		expect(p.peekMatches(TokenType.STRING_LITERAL)).toBe(false);
		expect(p.peekMatches(TokenType.STRING_LITERAL, "1")).toBe(false);

		p.next();
		expect(p.peekMatches(TokenType.INT_LITERAL, "1")).toBe(false);
		expect(p.peekMatches(TokenType.STRING_LITERAL, "hi")).toBe(true);

		p.next();
		expect(p.peekMatches(TokenType.STRING_LITERAL, "hi")).toBe(false);
	});

	test("PeekMatches with array", () => {
		const p = new TestableParser(new Lexer(new Source("1 'hi'")));
		expect(p.peekMatches(TokenType.INT_LITERAL, [])).toBe(false);
		expect(p.peekMatches(TokenType.INT_LITERAL, ["1"])).toBe(true);
		expect(p.peekMatches(TokenType.INT_LITERAL, ["1", "2"])).toBe(true);
		expect(p.peekMatches(TokenType.INT_LITERAL, ["2", "3"])).toBe(false);
	});
});

describe("Parser Errors", () => {
	test("Syntax errors", () => {
		expect(() => execLine("2 %")).toThrowError(JumpSyntaxError);
		expect(() => execLine("1 + 3 +")).toThrowError(JumpSyntaxError);
		expect(() => execLine("1 + 3 3")).toThrowError(JumpSyntaxError);
		expect(() => execLine("1 1")).toThrowError(JumpSyntaxError);
		expect(() => execLine("1 + 3 ^")).toThrowError(JumpSyntaxError);
		expect(() => execLine("(")).toThrowError(JumpSyntaxError);
		expect(() => execLine("(()")).toThrowError(JumpSyntaxError);
		expect(() => execLine("(()")).toThrowError(JumpSyntaxError);
		expect(() => execLine("ðŸ˜Ž")).toThrowError(JumpSyntaxError);
		expect(() => execLine("")).toThrowError(JumpSyntaxError);
	});
});
