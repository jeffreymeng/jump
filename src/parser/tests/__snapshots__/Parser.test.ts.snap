// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Internal next/peek/peekMatches Next Error Cases 1`] = `"Expected a '1' int_literal token, but got a '2' int_literal token."`;

exports[`Internal next/peek/peekMatches Next Error Cases 2`] = `"Unexpected end of input."`;
